Unitful literals (#3688)
modelica/ModelicaSpecification: chapters/unitexpressions.tex

The unit expression \lstinline!"T"! means tesla, but note that the letter \lstinline!T! is also the symbol for the prefix tera which has a multiplier value of 10\textsuperscript{12}.
\end{example}

\section{Unitful Literals}

The \lstinline[language=grammar]!numeric! in the grammar allows a unit to be attached to a numeric value.

> In this case, the unit is given as a \lstinline[language=grammar]!Q-IDENT! rather than a string, since the use of a string could be ambiguity interpreted as \lstinline[language=grammar]!description-string! in some contexts.

Reviewer: ```suggestion
In this case, the unit is given as a \lstinline[language=grammar]!Q-IDENT! rather than a string, since the use of a string could be ambiguously interpreted as \lstinline[language=grammar]!description-string! in some contexts.
```
To me "ambiguity interpreted" looks odd, and I prefer "ambiguously interpreted"; (possibly with "be" in the middle)

https://books.google.com/ngrams/graph?content=ambiguity+interpreted%2C+ambiguously+interpreted&year_start=1800&year_end=2022&corpus=en&smoothing=3

\begin{example}
An acceleration of 9.8 meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

\begin{example}
Adding 2~centimeters to a length.
A user has set \lstinline!x! to 10~centimeters with the use of \lstinline!displayUnit!, and intends to make \lstinline!y! 2~centimeters bigger than \lstinline!x!:
\begin{lstlisting}[language=modelica]
  parameter Real x(unit = "m", displayUnit = "cm") = 0.1;
equation

>   y = x + 2; /* Intention: Add 2 centimeters to x. */

Author: The difficulty is that we don't have a specification of how unit checking actually works, so in principle we can only speculate what this code actually does in different tools.  Explaining this in the comment would require too much space, so I gave the explanation in the example text below instead.

Regarding the style of comment, we have a mix of styles in the specification, and similar to different styles of whitespace around binary operators I think the current plan is to keep it this way.

\end{lstlisting}
The result is that 2~meters instead of 2~centimeters is being added to \lstinline!x!.

Using a unitful literal makes the intention clear:
\begin{lstlisting}[language=modelica]
  y = x + 2'cm';

> \end{lstlisting}

Reviewer: To make it clearer that it is just adding lines to clarify, and that the following lines are unchanged.

> \end{lstlisting}

Reviewer: It wasn't for the case above, or even for:
```
Real force(unit = "N") = 3;
Real distance(unit = "m") = 5;
Real lever(unit="m")=2;
Real work(unit = "J") = force * distance;
Real torque(unit="N.m") = force * lever;
```
but to _allow_ diagnostics for the subtly different:
```
Real torque2(unit="N.m") = work;
```
Similarly, for rotational frequencies.

> It is currently not expected that tools will perform an implicit unit conversion to make the addition possible, in which case the addition should be rejected as a unit inconsistency.

Reviewer: ```suggestion
Tools must infer the unit information of the unitful literal, including absolueValue when applicable, from the rest of the model, and verify that the given unit is compatible.
It is an error if the needed unit information cannot be inferred.
Diagnostics may be given for consistent units that shouldn't be mixed even (like \lstinline!"N.m"! and \lstinline!"J"!).
Special care is needed to handle rotational frequencies.
It is currently not expected that tools will perform an implicit unit conversion to make the addition possible, in which case the addition should be rejected as a unit inconsistency.
```
So, something like this.

The unit expression \lstinline!"T"! means tesla, but note that the letter \lstinline!T! is also the symbol for the prefix tera which has a multiplier value of 10\textsuperscript{12}.
\end{example}

\section{Unitful Literals}

The \lstinline[language=grammar]!numeric! in the grammar allows a unit to be attached to a numeric value.
In this case, the unit is given as a \lstinline[language=grammar]!Q-IDENT! rather than a string, since the use of a string could be ambiguously interpreted as \lstinline[language=grammar]!description-string! in some contexts.

\begin{example}
An acceleration of 9.8~meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

A unitful literal with a temperature unit represents an absolute value of temperature.

\begin{example}
The temperature \lstinline!20'degC'! converted to kelvin would be 293.15~kelvin, not 20~kelvin.
\end{example}

> 

Author: The old formulation has now been removed.  Resolving.

The unit expression \lstinline!"T"! means tesla, but note that the letter \lstinline!T! is also the symbol for the prefix tera which has a multiplier value of 10\textsuperscript{12}.
\end{example}

\section{Unitful Literals}

The \lstinline[language=grammar]!numeric! in the grammar allows a unit to be attached to a numeric value.
In this case, the unit is given as a \lstinline[language=grammar]!Q-IDENT! rather than a string, since the use of a string could be ambiguously interpreted as \lstinline[language=grammar]!description-string! in some contexts.

\begin{example}
An acceleration of 9.8~meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

> A unitful literal may only be used where unit consistency can be verified.

Author: Let's start over after all the discussions at the language group meeting.  Resolving.

The unit expression \lstinline!"T"! means tesla, but note that the letter \lstinline!T! is also the symbol for the prefix tera which has a multiplier value of 10\textsuperscript{12}.
\end{example}

\section{Unitful Literals}\label{unitful-literals}

The \productionref{unsigned-number} in the grammar allows a unit to be attached to a numeric value.
In this case, the unit is given as a \lexicalunitref{Q-IDENT} rather than a string, since the use of a string could be ambiguously interpreted as \productionref{description-string} in some contexts.

\begin{example}
An acceleration of 9.8~meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

When the modified component of a binding equation has a \lstinline!unit!-attribute, a binding equation consisting of a unitful literal or a negated unitful literal is implicitly converted to the component's unit.

\begin{nonnormative}
The predominant use of unscaled SI units in \lstinline!unit!-attributes in libraries (as this is a good way to avoid unit inconsistency without tool assistance to verify unit consistency) should be kept in mind when using unitful literals where there is no implicit unit conversion.

> That is, using units which are not equivalent to unscaled SI units for those unitful literals is likely to be a mistake.

Reviewer: We should use the traditional terminology from BIPM, especially as scaling is a bit odd for mass.

Something like:
```suggestion
The predominant use of coherent SI units in \lstinline!unit!-attributes in libraries (as this is a good way to avoid unit inconsistency without tool assistance to verify unit consistency) should be kept in mind when using unitful literals where there is no implicit unit conversion.
Coherent SI units are SI base units and coherent derived units.
Coherent indicate that they are without any scaling (e.g., including newton meter, but excluding centimeter), and ensures that the equations between the numerical values of quantities take exactly the same form as the equations between the quantities themselves.
Note that the SI base unit for mass is kilogram.
That is, using units which are not coherent units for those unitful literals is likely to be a mistake.
```
Part of a sentence is based on the BIPM quote:
> The word “coherent” here means that equations between the numerical values of quantities take exactly the same form as the equations between the quantities themselves.

The mistake can be detected by tools that perform (tool-dependent) analysis of unit consistency, while tools that don't analyze unit consistency are likely to just drop the unit and produce unexpected results, see example below.
\end{nonnormative}

\begin{example}
Implicit unit conversion in binding equations:
\begin{lstlisting}[language=modelica]
// Implicit unit conversion resolves inconsistent units:
Real x(unit = "m") = -5'cm';        // OK, same as -0.05'm'
// No implicit unit conversion since not (signed) literal:
Real y(unit = "m") = 0'cm' - 5'cm'; // Unit conflict
// No implicit unit conversion since not (signed) literal:
Real z(unit = "cm") = x;            // Unit conflict
// No implicit unit conversion since w has no unit-attribute:

> Real w = 5'cm';                     // OK

Reviewer: I would write something like:
```suggestion
Real w = 5'cm';                     // Not recommended, but accepted
```

Author: Not a big deal.  Applying.

The unit expression \lstinline!"T"! means tesla, but note that the letter \lstinline!T! is also the symbol for the prefix tera which has a multiplier value of 10\textsuperscript{12}.
\end{example}

\section{Unitful Literals}\label{unitful-literals}

The \productionref{unsigned-number} in the grammar allows a unit to be attached to a numeric value.
In this case, the unit is given as a \lexicalunitref{Q-IDENT} rather than a string, since the use of a string could be ambiguously interpreted as \productionref{description-string} in some contexts.

\begin{example}
An acceleration of 9.8~meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

When the modified component of a binding equation has a \lstinline!unit!-attribute, a binding equation consisting of a unitful literal or a negated unitful literal is implicitly converted to the component's unit.

\begin{nonnormative}

> The predominant use of coherent SI units in \lstinline!unit!-attributes in libraries (as this is a good way to avoid unit inconsistency without tool assistance to verify unit consistency) should be kept in mind when using unitful literals where there is no implicit unit conversion.

Author: We can even be kind and give the definition before explaining the consequences:

```suggestion
The predominant use of coherent SI units in \lstinline!unit!-attributes in libraries (as this is a good way to avoid unit inconsistency without tool assistance to verify unit consistency) should be kept in mind when using unitful literals where there is no implicit unit conversion.
A unit system is \emph{coherent} when every derived unit can be built from the base units purely by multiplying and dividing powers of them, with no numerical factor other than 1 anywhere in the construction.
```

Reviewer: Looks good.

Coherent SI units are SI base units and coherent derived units.

> Coherent indicate that they are without any scaling (e.g., including newton meter, but excluding centimeter), and ensures that the equations between the numerical values of quantities take exactly the same form as the equations between the quantities themselves.

Author: I would formulate it a bit differently to not make it sound as giving the definition of _coherent_:
```suggestion
A coherent system of units ensures that the equations between the numerical values of quantities take exactly the same form as the equations between the quantities themselves.
In particular, scaled SI units are not part of the coherent system, since, e.g., the product of centimeter and newton is not equivalent to joule.
```

Author: With the definition given above, it is better two swap the order:
```suggestion
For example, scaled SI units such as centimeter are not part of the coherent system, since they are derived units where the scaling relative to the base units is not 1.
A coherent system of units ensures that the equations between the numerical values of quantities take exactly the same form as the equations between the quantities themselves.
```

Author: Apply this as well, now that we have a clear definition right before this?

Author: I take that for a yes.

Note that the SI base unit for mass is kilogram.
That is, using units which are not coherent units for those unitful literals is likely to be a mistake.

> The mistake can be detected by tools that perform (tool-dependent) analysis of unit consistency, while tools that don't analyze unit consistency are likely to just drop the unit and produce unexpected results, see example below.

Reviewer: ```suggestion
The mistake can be detected by tools that perform (tool-dependent) analysis of unit consistency, while tools that don't analyze unit consistency are likely to just drop the unit and produce incorrect results, see example below.
```
I find that "unexpected" is a bit vague, and "incorrect" more clearly captures the issue.

\end{lstlisting}

Adding a unitful temperature literal to a relative temperature leads to the same ambiguity as when adding a temperature variable without \lstinline!absoluteValue! annotation.
That is, since \lstinline!20'degC'! could be either absolute or relative, a tool cannot infer the \lstinline!absoluteValue! status of the sum (compare \cref{modelica:absoluteValue}):
\begin{lstlisting}[language=modelica]
// Two variants with disallowed displayUnit conversion:
Real x(unit = "degC", displayUnit = "K") = tCRel + tC;
Real y(unit = "degC", displayUnit = "K") = tCRel + 20'degC';
\end{lstlisting}
\end{example}

\begin{example}
Adding 2~centimeters to a length.
A user has set \lstinline!x! to 10~centimeters with the use of \lstinline!displayUnit!, and intends to make \lstinline!y! 2~centimeters bigger than \lstinline!x!:
\begin{lstlisting}[language=modelica]
  parameter Real x(unit = "m", displayUnit = "cm") = 0.1;
equation
  y = x + 2; /* Intention: Add 2 centimeters to x. */
\end{lstlisting}
The result is that 2~meters instead of 2~centimeters is being added to \lstinline!x!.

Using a unitful literal makes the intention clear:
\begin{lstlisting}[language=modelica]

>   y = x + 2'cm';

Reviewer: ```suggestion
  y = x + 2'cm'; // Detectable error
```
We normally have comments on the lines of each incorrect line (even if it looks redundant), and don't rely on text succeeding or preceding it.

I'm not sure if `y = x + 2; /* Intention: Add 2 centimeters to x. */` could be clarified as well (it is a non-detectable error).

Author: I addressed this with the standard comment `Unit conflict` instead.  Please resolve if OK.

Reviewer: Ok

\begin{example}
An acceleration of 9.8~meters per square second can be expressed as \lstinline!9.8'm/s2'!.
\end{example}

When the modified component of a binding equation has a \lstinline!unit!-attribute, a binding equation consisting of a unitful literal or a negated unitful literal is implicitly converted to the component's unit.

\begin{example}
Implicit unit conversion in binding equations:
\begin{lstlisting}[language=modelica]
// Implicit unit conversion resolves inconsistent units:
Real x(unit = "m") = -5'cm';        // OK, same as -0.05'm'
// No implicit unit conversion since not (signed) literal:
Real y(unit = "m") = 0'cm' - 5'cm'; // Unit conflict
// No implicit unit conversion since not (signed) literal:
Real z(unit = "cm") = x;            // Unit conflict
// No implicit unit conversion since w has no unit-attribute:
Real w = 5'cm';                     // OK
\end{lstlisting}

There is no implicit unit conversion in equations:
\begin{lstlisting}[language=modelica]
  Real x(unit = "m");
equation
  x = -5'cm'    // Unit conflict
  x = -0.05'm'  // OK

> \end{lstlisting}

Reviewer: This was not what I thought we decided. We must have have misunderstood each other.
My impression was that these cases would be illegal, and only the case above would be handled.

I know that is a lot more restrictive, so that it completely postponed the problematic cases, instead of introducing them now and adding their semantic checks later.

Author: That sounds like a tremendous misunderstanding.  Without the ability to use unitful literals with consistent units in equations (just like when introducing a helper constant with the same unit), we can't start fixing unit issues in libraries.

The only problematic part about unitful literals is the implicit conversion, which we have carefully restricted so that it is only allowed where it is completely obvious what to do.
